File: app/src/main/java/ru/telematica/casco2go/service/http/HttpService.java
package ru.telematica.casco2go.service.http;

import android.text.TextUtils;
import android.util.Log;

import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Response;

import ru.telematica.casco2go.utils.Errors;

/**
 * Created by m.sidorov on 23.02.2018.
 */

public class HttpService {

    private static final String TAG = HttpService.class.getSimpleName();

    private final TelematicaApi apiService;

    private static final long SESSION_TIMEOUT = 2 * 60 * 60 * 1000;
    private String sessionId = "";
    private long sessionTime = 0;
    public boolean hasSessionId(){
        return TextUtils.isEmpty(sessionId) || System.currentTimeMillis() > sessionTime + SESSION_TIMEOUT;
    }
    public String getSessionId(){
        return sessionId;
    }


    public HttpService(TelematicaApi apiService){
        this.apiService = apiService;
    }

    public static String formatNetworkError(String methodName, Exception e){
        return String.format(String.format("http method [%s] - %s", methodName, Errors.getErrorMessage(e)));
    }

    public static class NetworkException extends RuntimeException {
        public NetworkException(String message){
            super(message);
        }
    }

    public static <T> void checkErrors(Response<T> response){
        if (!response.isSuccessful()){
            throw new NetworkException(String.format("network error [%d:%s]", response.code(), response.message()));
        }
        if (response.body() == null){
            throw new NetworkException("server json body is empty");
        }
    }

/*
    public Single<Track> getRadioActiveTrack(RadioEntity radio){
        return Single.just(radio)
                .subscribeOn(Schedulers.io())
                .map(radio1 -> {
                    if (radio1.isRestream()){
                        return new PlayerTrackInfo();
                    }
                    Response<PlayerTrackInfo> response = apiService.getPlayerInfo(radio1.getExtendedUrl()).execute();
                    checkErrors(response);
                    response.body().response_time = System.currentTimeMillis();
                    return response.body();
                })
                .map(playerTrackInfo -> playerTrackInfo.getTrack());
    }
*/

/*
    public SessionResponse getSession(){
        try {
            Response<SessionResponse> response = apiService.getSession(urls.getSession).execute();
            checkErrors(response);
            sessionId = response.body().session.rsid;
            sessionTime = System.currentTimeMillis();
            return response.body();
        } catch (Exception e){
            Log.e(TAG, formatNetworkError("getSession", e));
            throw new RuntimeException(formatNetworkError("getSession", e));
        }
    }
*/

}

File: app/src/main/java/ru/telematica/casco2go/service/http/HttpService.kt
package ru.telematica.casco2go.service.http

import android.util.Log
import io.reactivex.Single
import ru.telematica.casco2go.App
import ru.telematica.casco2go.model.AuthData
import ru.telematica.casco2go.model.ScoringData
import ru.telematica.casco2go.model.request.CreateTokenRequest
import ru.telematica.casco2go.model.request.RefreshTokenRequest
import ru.telematica.casco2go.model.response.*
import ru.telematica.casco2go.repository.ConfigRepository
import ru.telematica.casco2go.service.ScoringService
import ru.telematica.casco2go.utils.DateUtils
import ru.telematica.casco2go.utils.isNull
import java.util.*
import kotlin.collections.ArrayList

/**
 * Created by m.sidorov on 29.04.2018.
 */
class HttpService(val apiService: TelematicaApi) {

    private val TAG = HttpService::class.java.simpleName

    fun createAccessToken(request: CreateTokenRequest): CreateAccessTokenResponse {
        val response = apiService.createToken(request).execute()
        if (!response.isSuccessful || response.body() == null){
            Log.e(TAG, "error of createAccessToken(), errorCode = ${response.code()}")
            throw RuntimeException("Server auth response error, error code [${response.code()}]")
        }
        return response.body().isNull(CreateAccessTokenResponse())
    }

    fun refreshAccessToken(request: RefreshTokenRequest): CreateAccessTokenResponse {
        val response = apiService.refreshToken(request).execute()
        if (!response.isSuccessful || response.body() == null){
            Log.e(TAG, "error of createAccessToken(), errorCode = ${response.code()}")
            throw RuntimeException("Server auth response error, error code [${response.code()}]")
        }
        return response.body().isNull(CreateAccessTokenResponse())
    }

    private fun getAuthHeader(): String{
        if (!ConfigRepository.authData.isExpired()) {
            return ConfigRepository.authData.getAuthHeader()
        } else {
            return ""
        }
    }

    fun getTimeZone(latitude: Double, longtitude: Double): String {
        val response = apiService.getTimeZone(getAuthHeader(), latitude, longtitude).execute()
        if (!response.isSuccessful || response.body() == null || response.body()?.code != 1) {
            if (response.body() != null && response.body()?.code != 1) {
                Log.e(TAG, "error of getTimeZone(), code = ${response.body()?.code}")
                throw RuntimeException("Server getTimeZone response error, error code [${response.body()?.code}]")
            } else {
                Log.e(TAG, "error of getTimeZone(), errorCode = ${response.code()}")
                throw RuntimeException("Server getTimeZone response error, error code [${response.code()}]")
            }
        }

        val timeZone = response.body()?.timeZone.isNull("")
        return timeZone
    }

    fun startTrip(latitude: Double, longtitude: Double, maxDuraton: Int, timeZone: String): Single<StartTripResponse> {
        return Single.fromCallable {
            val response = apiService.startTrip(getAuthHeader(), latitude, longtitude, maxDuraton, timeZone, ConfigRepository.authData.sessionID).execute()
            if (!response.isSuccessful || response.body() == null || response.body()?.code != 1) {
                if (response.body() != null && response.body()?.code != 1) {
                    Log.e(TAG, "error of startTrip(), code = ${response.body()?.code}")
                    throw RuntimeException("Server startTrip response error, error code [${response.body()?.code}]")
                } else {
                    Log.e(TAG, "error of startTrip(), errorCode = ${response.code()}")
                    throw RuntimeException("Server startTrip response error, error code [${response.code()}]")
                }
            }

            response.body()
        }
    }

    fun finishTrip(): Single<ScoringData> {
        return Single.fromCallable {
            val response = apiService.finishTrip(getAuthHeader()).execute()
            if (!response.isSuccessful || response.body() == null || response.body()?.code != 1){
                if (response.body() != null && response.body()?.code != 1) {
                    Log.e(TAG, "error of startTrip(), code = ${response.body()?.code}")
                    throw RuntimeException("Server finishTrip response error, error code [${response.body()?.code}]")
                } else {
                    Log.e(TAG, "error of startTrip(), errorCode = ${response.code()}")
                    throw RuntimeException("Server finishTrip response error, error code [${response.code()}]")
                }
            }
            journeyToScoring(response.body()?.journey.isNull(Journey()))
        }
    }

    fun loadHistory(lastScoringData: ScoringData?): Single<List<ScoringData>>{
        var paramDateTime : String = lastScoringData?.startTimeS.isNull("")
        if (paramDateTime.isBlank()){
            paramDateTime = DateUtils.serverFormatter.formatDate(Date())
        }

        return Single.fromCallable{
            val response = apiService.loadHistory(getAuthHeader(), paramDateTime).execute()
            if (!response.isSuccessful || response.body() == null){
                Log.e(TAG, "error of loadHistory(), errorCode = ${response.code()}")
                throw RuntimeException("Server loadHistory response error, error code [${response.code()}]")
            }

            val result = ArrayList<ScoringData>()
            val list = response.body()?.data.isNull(ArrayList<Journey>())
            for (journey in list){
                result.add(journeyToScoring(journey))
            }
            result.apply {
                sortWith( compareBy { it.startTime } )
            }
        }
    }

    private fun journeyToScoring(journey: Journey): ScoringData {
        return ScoringData().apply {
            timeTripSec = journey.duration.isNull(0)
            timeInTrafficSec = journey.trafficJamDuration.isNull(0)
            timeInTravelSec = timeTripSec - timeInTrafficSec
            drivingLevel = journey.scorePercentKm.isNull(0f).toInt()
            val cost = ((timeTripSec / 60) * getOneMinuteCost()).toFloat()
            tripCost = cost - cost * getDiscount() / 100f

            startTimeS =journey.startTime
            if (!startTimeS.isNull("").isBlank()) {
                startTime = DateUtils.serverFormatter.parseDate(startTimeS.isNull(""))
            }
            if (!journey.finishTime.isNull("").isBlank()) {
                finishTime = DateUtils.serverFormatter.parseDate(journey.finishTime.isNull(""))
            }
            if (timeTripSec < ScoringService.MIN_TRIP_TIME_SEC) {
                drivingLevel = 0
            }
        }
    }

}